File: nodes/untokenizer/untokenizer.process.py
import tiktoken
from typing import Any, Dict, List, Union

async def process(inputs: Dict[str, Any], settings: Dict[str, Any], config: Dict[str, Any], nodeConfig: Dict[str, Any]) -> Dict[str, Any]:
    try:
        tokens = inputs.get('tokens')
        if not isinstance(tokens, list):
            raise Exception("Tokens input must be an array of numbers")

        # Get the tokenizer setting, default to cl200k_base
        tokenizer = settings.get('tokenizer', 'cl200k_base')

        # Validate tokenizer option
        if tokenizer not in ['cl200k_base', 'cl100k_base']:
            raise Exception(f"Unsupported tokenizer: {tokenizer}. Only cl200k_base and cl100k_base are supported.")

        # Get the encoding for the specified tokenizer
        # Python tiktoken uses different names
        encoding_name = 'cl200k_base' if tokenizer == 'cl200k_base' else 'cl100k_base'
        encoding = tiktoken.get_encoding(encoding_name)

        # Convert to regular list if needed
        token_list = list(tokens) if not isinstance(tokens, list) else tokens

        # Validate that all tokens are numbers
        for i, token in enumerate(token_list):
            if not isinstance(token, int):
                raise Exception(f"Token at index {i} is not a valid integer: {token}")

        # Decode the tokens back to text
        decoded_bytes = encoding.decode(token_list)
        text = decoded_bytes if isinstance(decoded_bytes, str) else decoded_bytes.decode('utf-8')

        return {
            "text": text
        }

    except Exception as error:
        print(f'Untokenizer error: {error}')
        raise Exception(f"Untokenizer error: {str(error)}")

File: nodes/untokenizer/untokenizer.process.js
import { encoding_for_model } from 'tiktoken';

export default async ({inputs, settings, config}) => {
    try {
        const tokens = inputs.tokens;
        if (!Array.isArray(tokens)) {
            throw new Error("Tokens input must be an array of numbers");
        }

        // Get the tokenizer setting, default to cl200k_base
        const tokenizer = settings.tokenizer || 'cl200k_base';

        // Validate tokenizer option
        if (!['cl200k_base', 'cl100k_base'].includes(tokenizer)) {
            throw new Error(`Unsupported tokenizer: ${tokenizer}. Only cl200k_base and cl100k_base are supported.`);
        }

        // Get the encoding for the specified tokenizer
        const encoding = encoding_for_model(tokenizer === 'cl200k_base' ? 'gpt-4o' : 'gpt-4');

        // Convert to regular array if it's a Uint32Array or similar
        const tokenArray = Array.isArray(tokens) ? tokens : Array.from(tokens);

        // Validate that all tokens are numbers
        for (let i = 0; i < tokenArray.length; i++) {
            if (typeof tokenArray[i] !== 'number' || !Number.isInteger(tokenArray[i])) {
                throw new Error(`Token at index ${i} is not a valid integer: ${tokenArray[i]}`);
            }
        }

        // Decode the tokens back to text
        const decodedBytes = encoding.decode(tokenArray);
        const text = typeof decodedBytes === 'string' ? decodedBytes : new TextDecoder().decode(decodedBytes);

        // Free the encoding to prevent memory leaks
        encoding.free();

        return {
            text: text
        };

    } catch (error) {
        console.log('Untokenizer error:', error);
        throw new Error(`Untokenizer error: ${error.message}`);
    }
};
